pallets/pallet-governance/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+
+			Ok(())
+		}
+
+		// Requires Root origin - sets the current authorities
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::force_rotate_authorities())]
+		pub fn force_rotate_authorities(
+			origin: OriginFor<T>,
+			new_members: Vec<T::AccountId>,
+		) -> DispatchResult {
+			// ensure root
+			ensure_root(origin)?;
+			Self::rotate_authorities(new_members);
+			Ok(())
+		}
+
+		// Requires Governance origin - sets the current authorities
+		#[pallet::call_index(3)]
+		#[pallet::weight(T::WeightInfo::council_rotate_authorities())]
+		pub fn council_rotate_authorities(
+			origin: OriginFor<T>,
+			new_members: Vec<T::AccountId>,
+		) -> DispatchResult {
+			// ensure governance
+			T::EnsureGovernance::ensure_origin(origin)?;
+
+			Self::rotate_authorities(new_members);
+			Ok(())
+		}
+	}
+
+	/// Custom governance origin
+	pub struct EnsureGovernance;
+
+	/// Implementation for EnsureOrigin trait for custom EnsureGovernance struct.
+	/// We use this to execute extrinsic by a governance origin.
+	impl<OuterOrigin> EnsureOrigin<OuterOrigin> for EnsureGovernance
+	where
+		OuterOrigin: Into<Result<RawOrigin, OuterOrigin>> + From<RawOrigin>,
+	{
+		type Success = ();
+
+		fn try_origin(o: OuterOrigin) -> Result<Self::Success, OuterOrigin> {
+			match o.into() {
+				Ok(o) => match o {
+					RawOrigin::GovernanceApproval => Ok(()),
+				},
+				Err(o) => Err(o),
+			}
+		}
+
+		#[cfg(feature = "runtime-benchmarks")]
+		fn try_successful_origin() -> Result<OuterOrigin, ()> {
+			Ok(RawOrigin::GovernanceApproval.into())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Dispatches a call from the governance origin, with transactional semantics, ie. if the
+		/// call dispatch returns `Err`, rolls back any storage updates.
+		fn dispatch_governance_call(
+			call: <T as Config>::RuntimeCall,
+		) -> DispatchResultWithPostInfo {
+			with_transaction(move || {
+				match call.dispatch_bypass_filter(RawOrigin::GovernanceApproval.into()) {
+					r @ Ok(_) => TransactionOutcome::Commit(r),
+					r @ Err(_) => TransactionOutcome::Rollback(r),
+				}
+			})
+		}
+
+		/// Increased proposal ID.
+		fn next_proposal_id() -> ProposalId {
+			NextProposalId::<T>::mutate(|id| {
+				*id = id.saturating_add(1);
+				*id
+			})
+		}
+
+		fn rotate_authorities(authorities: Vec<T::AccountId>) {
+			// Sets the new authority
+			let unique_members: BTreeSet<_> = authorities.into_iter().collect();
+			CurrentAuthorities::<T>::set(unique_members.clone());
+
+			// Reset all expiry blocks
+			if Expiry::<T>::iter().next().is_some() {
+				let expired_block = T::BlockNumberProvider::current_block_number()
+					.saturating_add(T::ExpiryPeriod::get());
+				// Initialize an empty set to store all elements
+				let mut all_elements: BTreeSet<ProposalId> = BTreeSet::new();
+
+				// Iterate through the storage map and collect all proposal ids.
+				for (_, expiry_set) in Expiry::<T>::iter() {
```

Comment:
dont need to check if it is empty.
you can use drain() here to take all data out